consensus/hotstuff/eventhandler/event_handler.go

```diff
@@ -376,15 +376,28 @@ func (e *EventHandler) processBlockForCurrentViewIfIsNextLeader(block *model.Blo
 	isNextLeader := true
 	curView := block.View
 
+	// before generating a vote, we first check that we are a valid committee
+	// member for this block. we may not be a valid committee member if we are
+	// joining in an upcoming epoch.
 	shouldVote := true
-	ownVote, err := e.voter.ProduceVoteIfVotable(block, curView)
-	if err != nil {
-		if !model.IsNoVoteError(err) {
-			// unknown error, exit the event loop
-			return fmt.Errorf("could not produce vote: %w", err)
-		}
-		log.Debug().Err(err).Msg("should not vote for this block")
+	_, err := e.committee.Identity(block.BlockID, e.committee.Self())
+	if errors.Is(err, model.ErrInvalidSigner) {
```

Comment:
❗❗  
We are trying to only have logic in the event handler that is for the core HotStuff. HotStuff is already complex enough as it is. Keeping intellectual complexity as possible (by strict modularization) is the only way we can keep the code maintainable. 

This logic is _not_ part of HotStuff. I feel it should go into the `voter`

Reply:
Moved to `Voter`